src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +121,169 @@ cdef class flint_poly(flint_elem):
+                raise ValueError("Number of variables does not equal number of names")
+        return nametup
+
+    @classmethod
+    def get_context(cls, slong nvars=1, ordering=Ordering.lex, names: Optional[str] = "x", nametup: Optional[tuple] = None):
+        """
+        Retrieve a context via the number of variables, `nvars`, the ordering, `ordering`, and either a variable
+        name string, `names`, or a tuple of variable names, `nametup`.
+        """
+
+        # A type hint of `ordering: Ordering` results in the error "TypeError: an integer is required" if a Ordering
+        # object is not provided. This is pretty obtuse so we check it's type ourselves
+        if not isinstance(ordering, Ordering):
+            raise TypeError(f"`ordering` ('{ordering}') is not an instance of flint.Ordering")
+
+        if nametup is not None:
+            key = nvars, ordering, nametup
+        elif nametup is None and names is not None:
+            key = nvars, ordering, cls.create_variable_names(nvars, names)
+        else:
+            raise ValueError("Must provide either `names` or `nametup`")
+
+        ctx = cls._ctx_cache.get(key)
+        if ctx is None:
+            ctx = cls._ctx_cache.setdefault(key, cls(*key))
+        return ctx
+
+    @classmethod
+    def from_context(cls, ctx: flint_mpoly_context):
+        return cls.get_context(
+            nvars=ctx.nvars(),
+            ordering=ctx.ordering(),
+            names=None,
+            nametup=ctx.names()
+        )
+
+
 cdef class flint_mpoly(flint_elem):
     """
     Base class for multivariate polynomials.
     """
 
+    def leading_coefficient(self):
+        return self.coefficient(0)
+
+    def to_dict(self):
+        return {self.exponent_tuple(i): self.coefficient(i) for i in range(len(self))}
+
+    def __contains__(self, x):
+        """
+        Returns True if `self` contains a term with exponent vector `x` and a non-zero coefficient.
+
+            >>> from flint import fmpq_mpoly_ctx, Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> p = ctx.from_dict({(0, 1): 2, (1, 1): 3})
+            >>> (1, 1) in p
+            True
+            >>> (5, 1) in p
+            False
+
+        """
+        return bool(self[x])
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def __pos__(self):
+        return self
+
+    def keys(self):
+        return self.monoms()
+
+    def values(self):
+        return self.coeffs()
+
+    def items(self):
+        """
+        Return the exponent vectors and coefficient of each term.
+
+            >>> from flint import fmpq_mpoly_ctx, Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> list(f.items())
+            [((1, 1), 4), ((1, 0), 2), ((0, 1), 3), ((0, 0), 1)]
+
+        """
+        return zip(self.keys(), self.values())
```

Comment:
Maybe these should just be removed now.

Reply:
Now that we have `monoms` and `coeffs` which perform the same role.

Reply:
Agreed, but I will note here that `.items` is slightly different that the `.terms` currently implemented (and sympy), it returns a list of the terms of the polynomials rather than a tuple of exponent tuple and coefficient. (I swear I commented about that but I cannot find it)